### PR TITLE
Make godeps reproducible

### PR DESCRIPTION
**Pull request by eparis**
This institutes 3 new concepts.
1. hack/godeps/restore.sh  -- which is actually able to get close to succeeding restoring godeps
2. hack/godeps/patches  -- a series of patches which we 'pick' on top of 'clean' godeps
3. hack/godeps/save.sh -- which actually takes the results of GOPATH, applies the patches, and commits it (with `-c`) to the repo.

This PR then uses those tools to regenerate godeps from scratch. Which is turn means we delete almost 750,000 lines from the repo.

### CODE REVIEW

**Comment by eparis** [comment]
@liggitt @ironcladlou @deads2k @smarterclayton 

Anything that removes 750,000 lines from the repo must be good, right?

**Comment by eparis** [comment]
I'm open to the idea that item 2 should be a list of previously committed shas. So to verify reproducibility one would generate clean godeps then cherry-pick those previous shas.  Instead of having actual patches in /patches/.  Having actual patches in /patches/ really does make it easy to see how much debt we've accumulated.  (less than I expected)

**Comment by deads2k** [comment]
> Anything that removes 750,000 lines from the repo must be good, right?

Why did we lose test files? For instance, package ` Godeps/_workspace/src/github.com/coreos/go-etcd/etcd` is still present, but we can no longer run its unit tests.

**Comment by eparis** [comment]
do you run it now?  because godep doesn't include 'upstream' _test.go anymore. It's easy to put those back (but going to hurt my 750k number)

**Comment by deads2k** [comment]
> do you run it now? because godep doesn't include 'upstream' _test.go anymore. It's easy to put those back (but going to hurt my 750k number)

Not on any regular basis (except for kube), but I'd like to think that anyone contemplating a patch would run the associated unit tests.  Excluding myself of course, I'm a special snowflake.

**Comment by smarterclayton** [comment]
If we formalize our forks for carries we can run the tests on the fork.

**Comment by eparis** [comment]
Given the results of the coreos upstream PR, I guess you are a special little snowflake   :)

It is obvious how few forks we have. docker/distribution seems to be the only one that might really make sense .

File: hack/godeps/save.sh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+# build the godep tool
+go get -u github.com/tools/godep 2>/dev/null
+GODEP="${GOPATH}/bin/godep"
+
+# Use to following if we ever need to pin godep to a specific version again
+#pin-godep 'v53'
+
+rm -rf Godeps
+"${GODEP}" save ./...
```
Comment (liggitt) [inline hack/godeps/save.sh]: we want test files from kube
we also want to run `hack/copy-kube-artifacts.sh` after saving to pull in more kube stuff

File: hack/godeps/restore.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+preload-dep() {
+  org="$1"
+  project="$2"
+  sha="$3"
+
+  # Go get stinks, which is why we whave the || true...
+  go get -d "${org}/${project}" >/dev/null 2>&1 || true
+  repo_dir="${GOPATH}/src/${org}/${project}"
+  pushd "${repo_dir}" > /dev/null
+    git remote update > /dev/null
```
Comment (liggitt) [inline hack/godeps/restore.sh]: people name their remotes different things... which remotes will this update?

File: hack/godeps/restore.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  local orig_project="$2"
+  local alt_org="$3"
+  local alt_project="$4"
+
+  # Go get stinks, which is why we whave the || true...
+  go get -d "${orig_org}/${orig_project}" >/dev/null 2>&1 || true
+
+  repo_dir="${GOPATH}/src/${orig_org}/${orig_project}"
+  pushd "${repo_dir}" > /dev/null
+    git remote add "${alt_org}-remote" "https://${alt_org}/${alt_project}.git" > /dev/null
```
Comment (liggitt) [inline hack/godeps/restore.sh]: not all our remotes are git... I know of at least one mercurial repo

File: hack/godeps/restore.sh (hunk 2 above)
Comment (eparis) [inline hack/godeps/restore.sh]: all of them

File: hack/godeps/restore.sh (hunk 3 above)
Comment (eparis) [inline hack/godeps/restore.sh]: thankfully we aren't picking on top of an hg repo, so it isn't needed yet...

File: hack/godeps/save.sh (hunk 1 above)
Comment (eparis) [inline hack/godeps/save.sh]: will fix

**Comment by liggitt** [comment]
why wouldn't we make the tooling find the carried patch commits and apply them in order to verify, instead of making developers keep double books?

**Comment by eparis** [comment]
generating the pick list instead of explicitly doing a pick list is a good idea. i'm not sure how to do it, but can think on it for sure...

**Comment by danmcp** [comment]
@eparis Still valid?  Next steps?

**Comment by eparis** [comment]
It was, but obviously needs some rebase-in. I'll try to get it back soon.

**Comment by soltysh** [comment]
We currently have `hack/godep-restore.sh` and `hack/godep-save.sh`. If anything, I'd like to see us running the former as a verification step.

**Comment by soltysh** [comment]
I think we can close this, since I've introduced something along this way in https://github.com/openshift/origin/pull/12011, where one of the commits runs `hack/godep-restore.sh` to verify if newly modified deps are restorable.

**Pull request closed without merging**